### PR TITLE
functest: If doc changes only, then skip functests

### PR DESCRIPTION
**Pull request by fabiand**
If there are only docs or ownership changes, then there were no code changes, thus why should we run functests in that case?

This is a pragmatic change in order to exit early in case of only known non-CI impacting changes.

### What this PR does
Before this PR: functests were run on every change

After this PR: functests are exited early ("not run") in the case of changes limited to `docs/*` and `*/OWNERS`.

### Why we need it and why it was done in this way
The following tradeoffs were made: Simple change in order to safe compute resources

The following alternatives were considered: Adding it to prow, but this seemed to be complex.

Links to places where the discussion took place: 
- None

### Special notes for your reviewer

None

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [x] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [x] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

**Comment by fabiand** [comment]
@dhiller @brianmcarey thoughts?
@xpivarc @acardace ?

**Comment by 0xFelix** [comment]
Do you deem it to complex to add to Prow because it needs to be added to every job?

**Comment by fabiand** [comment]
> Do you deem it to complex to add to Prow because it needs to be added to every job?

@dhiller is the expert, my understanding is that the combination of mandatory lanes and skipping them makes it complicated.

Thus in this PR, all the lane configuration stays, and KubeVirt CI will be brought up, but at least we then don't progress further.

**Comment by xpivarc** [comment]
I think this is not worth it most of the time. Do you have any numbers to support this?

**Comment by dhiller** [comment]
> > Do you deem it to complex to add to Prow because it needs to be added to every job?
> 
> @dhiller is the expert, my understanding is that the combination of mandatory lanes and skipping them makes it complicated.

At first we were looking at using either `run_if_changed` or `skip_if_changed` job configuration options. However they don't work well together with branch_protector, since they make the jobs conditional and only non conditional jobs can be required (aka merge gating).

We then thought about forking the trigger plugin, since the best way to do it would be to just not run the jobs at all and set the check to green directly. But that seemed to be too much work, since we then would have to maintain all the test-infra stuff ourselves. (Reminder to myself: I should ask Petr Muller whether he sees an easy way to do this...)

And we finally thought about a similar thing that Fabian does in this PR (a bit more fancy though, I wonder whether Or Shovals PR still exists?), but we couldn't find a consensus on that matter.

(Sorry,bad internet ATM, will update with links when I can)

> 
> Thus in this PR, all the lane configuration stays, and KubeVirt CI will be brought up, but at least we then don't progress further.

**Comment by EdDev** [comment]
@oshoval tried to suggest it.

https://github.com/kubevirt/kubevirt/pull/10019

**Comment by fabiand** [comment]
Thanks  @EdDev 

@oshoval howdy. How does your PR compare to this one. Whic would you pick?


What I like about mine: I don't add the remote :) However, I don't know if mine is working tho.

**Comment by fabiand** [comment]
@xpivarc please see Daniel's remark. Thus due to the lack of a realistic alternative this PR got proposed.

**Comment by oshoval** [comment]
> @oshoval howdy. How does your PR compare to this one. Whic would you pick?
> 
> What I like about mine: I don't add the remote :) However, I don't know if mine is working tho.

Hi Fabian
Thanks for the PR, once this PR is tested it is great

**Comment by fabiand** [comment]
/hold

WIP

File: automation/test.sh
Comment (oshoval) [inline automation/test.sh:38]: i think this one isnt needed

File: automation/test.sh
Comment (oshoval) [inline automation/test.sh:35]: maybe we can simplify to just check if the `diff | grep -v` is empty please ?

File: automation/test.sh
Comment (oshoval) [inline automation/test.sh:43]: dont think we need this on the long run, wdyt ?

Comment (fabiand) [reply automation/test.sh]: Agreed

File: automation/test.sh
Comment (fabiand) [reply automation/test.sh]: Ack

**Comment by fabiand** [comment]
@oshoval I'm now happy with this PR.

Please review.

**Comment by fabiand** [comment]
@dhiller @brianmcarey FYI, your feedback is welcome as well.

File: automation/test.sh
Comment (oshoval) [reply automation/test.sh]: reduces number of variables, simpler imho

diff=$(... | grep -v markdowns)
if diff == empty list; exit 0; fi

**Comment by fabiand** [comment]
@xpivarc the data: For PRs matching the pattern, the runtime is reduced from N x 10min+ to N x 30s.

If we assume N = 15 (grep e2e | wc -l), then you have:
Before: 150min
After: 8min

Time savings for those PRs: 18x

HTH

File: automation/test.sh
Comment (oshoval) [inline automation/test.sh:33]: Some more to consider skipping please png, svg, rst, txt, LICENSE, NOTICE, SECURITY_CONTACTS, PATENTS, MAINTAINERS
(we have some on non docs folder)

nit: consider please adding $ to .md, we dont have a file with .md in the middle, so we should be save anyhow
unless we can expect something weird like that

File: automation/test.sh
Comment (fabiand) [reply automation/test.sh]: Ok, fair!

File: automation/test.sh
Comment (fabiand) [reply automation/test.sh:33]: +1 on txt
Generally I'd like to limit the expression to files whicih we expect to changes
Owners, docs.
License, Notice, Security_contacts ... maybe the change some day, but not regularly.

Thus: I'm keeping the list smaller in favor of readbaility.

Technically however you are correct.

**Comment by acardace** [comment]
/lgtm

@fabiand can you just change the PR title? Maybe to something more informative like 'don't run the CI if there are no code changes'

**Comment by acardace** [comment]
/unhold

@fabiand I think you want to unhold as well right?

**Review by EdDev — COMMENTED** [review]
Very nice.
Can we have this tested on a temp PR that rebases on it and applies a doc change?

**Comment by fabiand** [comment]
Neat idea!

#11164 is based on this one and touches a docs file.

**Comment by fabiand** [comment]
Darn. It will not work, @EdDev , because it also includes the test change. Well, lemme hack it up.

File: automation/test.sh
Comment (brianmcarey) [inline automation/test.sh:33]: `PULL_PULL_SHA` does not exist in batch jobs so we would have to handle the batch case too or else this could end up quietly skipping all the job in a batch.

Also adding a check to make sure `CI_GIT_ALL_CHANGES` is not empty would be good in case one of the SHA env variables is not present like in the batch case.

**Comment by dhiller** [comment]
> > > Do you deem it to complex to add to Prow because it needs to be added to every job?
> > 
> > 
> > @dhiller is the expert, my understanding is that the combination of mandatory lanes and skipping them makes it complicated.
> 
> At first we were looking at using either `run_if_changed` or `skip_if_changed` job configuration options. However they don't work well together with branch_protector, since they make the jobs conditional and only non conditional jobs can be required (aka merge gating).
> 
> We then thought about forking the trigger plugin, since the best way to do it would be to just not run the jobs at all and set the check to green directly. But that seemed to be too much work, since we then would have to maintain all the test-infra stuff ourselves. (Reminder to myself: I should ask Petr Muller whether he sees an easy way to do this...)
> 
> And we finally thought about a similar thing that Fabian does in this PR (a bit more fancy though, I wonder whether Or Shovals PR still exists?), but we couldn't find a consensus on that matter.
> 
> (Sorry,bad internet ATM, will update with links when I can)
> 
> > Thus in this PR, all the lane configuration stays, and KubeVirt CI will be brought up, but at least we then don't progress further.

So I asked Petr Muller on k8s prow Slack and this was his response: https://kubernetes.slack.com/archives/CDECRSC5U/p1707234214828279?thread_ts=1707150831.085229&cid=CDECRSC5U

So I think we have a misconfiguration somewhere, and we should check how we can fix this and get it working instead of merging this PR. WDYT?

**Comment by oshoval** [comment]
> So I asked Petr Muller on k8s prow Slack and this was his response: https://kubernetes.slack.com/archives/CDECRSC5U/p1707234214828279?thread_ts=1707150831.085229&cid=CDECRSC5U
> 
> So I think we have a misconfiguration somewhere, and we should check how we can fix this and get it working instead of merging this PR. WDYT?

If we use `skip_if_only_changed`, we wont able to act to those jobs as phase 2, as this field must be empty 
for `require_manually_triggered_jobs`.
I think this PR once we cover all the required corner cases is standalone, simple and good enough to have.

File: automation/test.sh
Comment (dhiller) [reply automation/test.sh:33]: Oh, good catch!

@fabiand FYI

Comment (EdDev) [reply automation/test.sh:33]: Sounds like we should not do this "skip" for batch runs, i.e. to condition it on the value existence of both env vars.

**Comment by dhiller** [comment]
> @xpivarc the data: For PRs matching the pattern, the runtime is reduced from N x 10min+ to N x 30s.
> 
> If we assume N = 15 (grep e2e | wc -l), then you have: Before: 150min After: 8min
> 
> Time savings for those PRs: 18x
> 
> HTH

Makes sense. However, I have the feeling that @xpivarc wanted to point out that in relation to all the PRs we have, not that many are docs only or OWNERS only PRs.

If this is the case that for say 1/20 PRs only we shave off that time, it might be getting marginal. IIUC your logic correctly then the difference would be around a reduction of 5% for the runtime of all the 20 PRs.

Like `(20*150 - 19*150+8)/(20*150) ~= 5.2%` of total runtime reduction.

It's still worth it to shave this off, since it's resource usage, and this simple construct should be easy enough to maintain.

File: automation/test.sh
Comment (dhiller) [reply automation/test.sh:33]: I think the check will take care of that itself, since the change set reflects whether it only contains docs in the merged PR, no?

**Comment by fabiand** [comment]
I've refreshed the PR. #11164 (is also updated) shows that it fails fast.

File: automation/test.sh
Comment (fabiand) [reply automation/test.sh:30]: Thanks, I did consider it, and I would liket o keep the comment, and I would like to keep it not in a function - mainly to align with the rest of the doc, and I don't see how it can be reused.

File: automation/test.sh
Comment (fabiand) [reply automation/test.sh:34]: Ack, fixed.

**Comment by oshoval** [comment]
Thanks

/lgtm

**Comment by fabiand** [comment]
Thanks!

@dhiller any further feedback or good?

File: automation/test.sh
Comment (dhiller) [reply automation/test.sh:34]: FTR here's the docs for this: https://docs.prow.k8s.io/docs/jobs/#job-environment-variables

File: automation/test.sh
Comment (dhiller) [inline automation/test.sh:33]: So I checked the script with a change in `README.md` which it didn't catch. So I believe that the original expression only works for all OWNERS/ALIASES, txt and md files in subdirectories, but not for the root level.

```suggestion
    SKIP_PATTERN="^((docs)/.*|.*(OWNERS|OWNERS_ALIASES|.*\.(md|txt)))$"
```

File: automation/test.sh
Comment (dhiller) [inline automation/test.sh:42]: nit: not directly related, but IIUC this should be in double braces and is not valid (at least shellcheck complained.
```suggestion
if [[ ${CI} == "true" ]]; then
```

**Review by dhiller — COMMENTED** [review]
@fabiand I have one comment and a nit

File: automation/test.sh
Comment (dhiller) [reply automation/test.sh:33]: This one is not perfectly valid either TBH, but I just wanted to point out what I meant.

Comment (dhiller) [reply automation/test.sh:33]: That might be better:

```suggestion
    SKIP_PATTERN="^((docs)/.*|(.*/)*(OWNERS|OWNERS_ALIASES|.*\.(md|txt)))$"
```

File: automation/test.sh
Comment (oshoval) [reply automation/test.sh:42]: you are correct that otherwise CI should be in quotes

File: automation/test.sh
Comment (oshoval) [reply automation/test.sh:33]: nice
we can also take `basename` and then we just mess with the filename wdyt?

Comment (fabiand) [reply automation/test.sh:33]: I've reworked the expression.

File: automation/test.sh
Comment (fabiand) [reply automation/test.sh:42]: `[` is usually a binary. `[[` is part of the `bash` syntax.

File: automation/test.sh
Comment (oshoval) [inline automation/test.sh:33]: now it skips only OWNERS (or OWNERS_ALIASES) on root isnt it ?
and we have more on non vendor, i.e 
```
./bazel/OWNERS
./cmd/OWNERS
./tests/network/OWNERS
./tests/storage/OWNERS
./tests/OWNERS
./tests/libvmi/OWNERS
./tests/performance/OWNERS
./staging/src/kubevirt.io/api/OWNERS
./OWNERS
./tools/OWNERS
./pkg/virt-operator/resource/OWNERS
./pkg/hotplug-disk/OWNERS
./pkg/network/OWNERS
./pkg/storage/OWNERS
./pkg/monitoring/OWNERS
./hack/OWNERS
```

Comment (fabiand) [reply automation/test.sh:33]: No, it should skip on OWNERS and OWNERS_ALIASES everywhere, which is correct.

**Review by dhiller — APPROVED** [review]
@fabiand nice, looks good to me!

TIL that `|` has higher precedence than `[^$]`, thanks!

/approve

**Comment by xpivarc** [comment]
> > @xpivarc the data: For PRs matching the pattern, the runtime is reduced from N x 10min+ to N x 30s.
> > If we assume N = 15 (grep e2e | wc -l), then you have: Before: 150min After: 8min
> > Time savings for those PRs: 18x
> > HTH
> 
> Makes sense. However, I have the feeling that @xpivarc wanted to point out that in relation to all the PRs we have, not that many are docs only or OWNERS only PRs.
> 
> If this is the case that for say 1/20 PRs only we shave off that time, it might be getting marginal. IIUC your logic correctly then the difference would be around a reduction of 5% for the runtime of all the 20 PRs.
> 
> Like `(20*150 - 19*150+8)/(20*150) ~= 5.2%` of total runtime reduction.
> 
> It's still worth it to shave this off, since it's resource usage, and this simple construct should be easy enough to maintain.

Well, those 19 PRs would probably have more revisions and that would increase their runtimes even more, making the one PR even more negligible.

**Pull request merged**
